packages/shaderlab/src/shaders/shadingPBR/surfaceDataPBR.glsl

```diff
@@ -0,0 +1,181 @@
+                discard;
+            }
+        #endif
+
+        #ifdef MATERIAL_HAS_ROUGHNESS_METALLIC_TEXTURE
+            vec4 metalRoughMapColor = texture2D( material_RoughnessMetallicTexture, v.v_uv );
+            roughness *= metalRoughMapColor.g;
+            metal *= metalRoughMapColor.b;
+        #endif
+
+        #ifdef MATERIAL_HAS_SPECULAR_GLOSSINESS_TEXTURE
+            vec4 specularGlossinessColor = texture2D(material_SpecularGlossinessTexture, v.v_uv );
+            #ifndef ENGINE_IS_COLORSPACE_GAMMA
+                specularGlossinessColor = gammaToLinear(specularGlossinessColor);
+            #endif
+            specularColor *= specularGlossinessColor.rgb;
+            glossiness *= specularGlossinessColor.a;
+        #endif
+
+
+        #ifdef IS_METALLIC_WORKFLOW
+            surfaceData.diffuseColor = baseColor.rgb * ( 1.0 - metal );
+            surfaceData.specularColor = mix( vec3(f0), baseColor.rgb, metal );
+            surfaceData.roughness = roughness;
+        #else
+            float specularStrength = max( max( specularColor.r, specularColor.g ), specularColor.b );
+            surfaceData.diffuseColor = baseColor.rgb * ( 1.0 - specularStrength );
+            surfaceData.specularColor = specularColor;
+            surfaceData.roughness = 1.0 - glossiness;
+        #endif
+
+        surfaceData.roughness = max(surfaceData.roughness, getAARoughnessFactor(surfaceData.normal));
+
+        #ifdef MATERIAL_ENABLE_CLEAR_COAT
+            surfaceData.clearCoat = material_ClearCoat;
+            surfaceData.clearCoatRoughness = material_ClearCoatRoughness;
+            #ifdef MATERIAL_HAS_CLEAR_COAT_TEXTURE
+                surfaceData.clearCoat *= (texture2D( material_ClearCoatTexture, v.v_uv )).r;
+            #endif
+            #ifdef MATERIAL_HAS_CLEAR_COAT_ROUGHNESS_TEXTURE
+                surfaceData.clearCoatRoughness *= (texture2D( material_ClearCoatRoughnessTexture, v.v_uv )).g;
+            #endif
+            surfaceData.clearCoat = saturate( surfaceData.clearCoat );
+            surfaceData.clearCoatRoughness = max(surfaceData.clearCoatRoughness, getAARoughnessFactor(surfaceData.clearCoatNormal));
+        #endif
+
+        #ifdef MATERIAL_IS_TRANSPARENT
+            surfaceData.opacity = baseColor.a;
+        #else
+            surfaceData.opacity = 1.0;
+        #endif
+
+        vec3 emissiveRadiance = material_EmissiveColor;
+        #ifdef MATERIAL_HAS_EMISSIVETEXTURE
+            vec4 emissiveColor = texture2D(material_EmissiveTexture, v.v_uv);
+            #ifndef ENGINE_IS_COLORSPACE_GAMMA
+                emissiveColor = gammaToLinear(emissiveColor);
+            #endif
+        emissiveRadiance *= emissiveColor.rgb;
+        #endif
+        surfaceData.emissive = emissiveRadiance;
+
+}
+
+void initSurfaceData(Temp_Varyings v, out SurfaceData surfaceData, bool isFrontFacing){
+    initGeometry(v, surfaceData, isFrontFacing);
+    initMaterial(v, surfaceData);
+}
+
+
+#endif
```

Comment:
Weird abstraction, class named `surfaceDataPBR.glsl` include `initMaterial` and ....

Reply:
done